datafusion/expr/src/planner.rs

```diff
@@ -161,6 +162,28 @@ pub trait ExprPlanner: Send + Sync {
     ) -> Result<PlannerResult<Vec<Expr>>> {
         Ok(PlannerResult::Original(args))
     }
+
+    /// Plans a `RawAggregateUDF` based on the given input expressions.
+    ///
+    /// Returns a `PlannerResult` containing either the planned aggregate function or the original
+    /// input expressions if planning is not possible.
+    fn plan_aggregate_udf(
+        &self,
+        aggregate_function: RawAggregateUDF,
+    ) -> Result<PlannerResult<RawAggregateUDF>> {
+        Ok(PlannerResult::Original(aggregate_function))
```

Comment:
If we agree to convert `count()` to `count(1)` in planner. The conversion would be like

```rust
    fn plan_aggregate_udf(
        &self,
        aggregate_function: RawAggregateUDF,
    ) -> datafusion_common::Result<PlannerResult<RawAggregateUDF>> {

        let RawAggregateUDF { udf, args, distinct, filter, order_by, null_treatment } = aggregate_function;
        if udf.name() == "count" && args.is_empty() {
            let args = vec![lit(1)];
            let expr = Expr::AggregateFunction(AggregateFunction {func_def: datafusion_expr::expr::AggregateFunctionDefinition::UDF(udf), args, distinct, filter, order_by, null_treatment});
            return Ok(PlannerResult::Planned(expr))
        }

        let aggregate_function = RawAggregateUDF { udf, args, distinct, filter, order_by, null_treatment };
        Ok(PlannerResult::Original(aggregate_function))
    }
```

Reply:
I'm not sure what the path is after that to getting from `count(1)` to `count()` as we're talking about in https://github.com/apache/datafusion/pull/11229 (that also works in `.select`). I think until there's a clear path to that, I'm hesitant to expand the scope of this PR.

Reply:
I think the conversion from `count(1)` to `count()` is more like an optinization (rather than something for hte sql planner) as it would apply equally to SQL and to dataframe apis

Perhaps we could add it as part of https://docs.rs/datafusion/latest/datafusion/logical_expr/trait.AggregateUDFImpl.html#method.simplify 🤔

Reply:
> For sql, count(), count(1) and count(*) are equivalent things that can all to one single Expr, so I think it is possible to have the conversion in planner.

I agree they are equivalent things. But I am thinking it may be hard to teach everyone who builds `Exprs` to call `count_star` rather than `count(lit(1))` or `count(Expr::Wildcard)`

To be clear I don't have a massively strong opinion, but it seems like since we don't control people's creation of `Expr` we aren't going to be able to prevent stuff like `count(1)`

Reply:
The rewrite could be simple enough that we don't need to push down to optimizer.

```rust
pub fn count_star() -> Expr {
    Expr::AggregateFunction(datafusion_expr::expr::AggregateFunction::new_udf(
        count_udaf(),
        vec![Expr::Literal(COUNT_STAR_EXPANSION)],
        false,
        None,
        None,
        None,
    ))
}

pub fn count(expr: Expr) -> Expr {
    // For backward compatility, could use count_star instead
    if let Expr::Wildcard { qualifier: _ } = expr {
        count_star()
    } else {
        Expr::AggregateFunction(datafusion_expr::expr::AggregateFunction::new_udf(
            count_udaf(),
            vec![expr],
            false,
            None,
            None,
            None,
        ))
    }
}
```

I think we need a clear role between expression rewrite in planner vs rewrite in optimizer. 
ExprPlanner is the first spot we get the expressions from SQLExpr and build up datafusion::Expr. For syntax rewrite like dict `{a: b}`, compound id `a.b` or equivalent args rewrite `count(*)` are all good candidate to determine the expected datafusion::Expr in ExprPlanner. Others should push down to optimizer

If the reason to rewrite expression in optimizer is because that benefit both sql and dataframe API, maybe redesign dataframe API is what we need 🤔 ?

Reply:
>
> I think we need a clear role between expression rewrite in planner vs rewrite in optimizer. ExprPlanner is the first spot we get the expressions from SQLExpr and build up datafusion::Expr. For syntax rewrite like dict `{a: b}`, compound id `a.b` or equivalent args rewrite `count(*)` are all good candidate to determine the expected datafusion::Expr in ExprPlanner. Others should push down to optimizer since they may need const evaluation or multiple passes.
> 
> If the reason to rewrite expression in optimizer is because that benefit both sql and dataframe API, maybe redesign dataframe API is what we need 🤔 ?

I guess in my mind the distinction is 
1.  the planner is a mechanical transformation to something that could be run 
2. the optimizer rewrites expressions / plans to keep the exact same semantics (output results) but faster.

So in my mind the transformation of `count(1)` --> `count()` is an optimization as it potentially avoids having to accumulate a known column in the execution. Maybe there is a better way to think about it

This might be a good distinction to clarify in the documentation 🤔

Reply:
> I guess in my mind the distinction is
the planner is a mechanical transformation to something that could be run
the optimizer rewrites expressions / plans to keep the exact same semantics (output results) but faster.

I agree with this. Rewrite in logical optimizer results to the better one.

> So in my mind the transformation of count(1) --> count() is an optimization as it potentially avoids having to accumulate a known column in the execution. Maybe there is a better way to think about it

In the case of count, `count(*)`, `count(1)`, and `count()` are equivalent things, not one is faster than the other.

The optimization that counts the row number is not in the logical optimizer but in the physical optimizer. Therefore, the transformation from `count()` and `count(*)` to `count(1)` is not close to the second point IMO. What we are doing before the physical optimizer is to transform the equivalent expr to a single one (`count(1)`), so we just need to process that one in the physical optimizer. Note that we could process those 3 expressions individually in the physical optimizer, but standardizing to a single expression as early as possible might reduce the complexity of the downstream query processing. *It is more like rewriting expressions to the equivalent expression* to me.

> This might be a good distinction to clarify in the documentation 🤔

It is nice to do so, we should clarify the role between these two.